### PR TITLE
Show / hide via api only (configurable)

### PR DESCRIPTION
**Pull request by piotr-dajlido**

### CODE REVIEW

**Comment by vlio20** [comment]
Didn't went through all of your changes, I see there are no test and no documentation were added. Plus, could provide some use case for this feature?

**Comment by piotr-dajlido** [comment]
Sure, use case is simple, more control over component behavior which collided with my business requirements, that's the entire reason for this fork. I have some actions bonded to focus and click events which modifiy the view and poping out calendar was rather annoying, while I still wanted to have it linked by FormControl where you can type date by hand. I can provide tests and docs as long as you agree for this kind of functionality, otherwise could you give me some hint how to build this project into npm package (index.js etc..?)

**Comment by vlio20** [comment]
Sure! Go ahead! I will be happy to merge. Please make sure this functionality is working for all type of pickers.

File: src/app/date-picker/date-picker-config.model.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@ export interface IDatePickerConfig extends IDayCalendarConfig,
                                            ITimeSelectConfig {
   closeOnSelect?: boolean;
   closeOnSelectDelay?: number;
+  apiControlOnly?: boolean;
```
Comment (vlio20) [inline src/app/date-picker/date-picker-config.model.ts]: should be available in all date type components. Also, the name is a bit problematic as it currently only implies on open and close apis. So I would give it a more specific naming.

File: src/app/date-picker/date-picker.component.ts [hunk 2]
```diff
@@ -69,11 +69,11 @@ import {Moment, unitOfTime} from 'moment';
   ]
 })
 export class DatePickerComponent implements OnChanges,
-                                            OnInit,
-                                            AfterViewInit,
-                                            ControlValueAccessor,
-                                            Validator,
-                                            OnDestroy {
+  OnInit,
```
Comment (vlio20) [inline src/app/date-picker/date-picker.component.ts]: please change it to the same alignment that was set here.

File: src/app/date-picker/date-picker.component.ts [hunk 3]
```diff
@@ -112,7 +112,7 @@ export class DatePickerComponent implements OnChanges,
   popupElem: HTMLElement;
   handleInnerElementClickUnlisteners: Function[] = [];
   globalListnersUnlisteners: Function[] = [];
-  validateFn: (inputVal: CalendarValue) => {[key: string]: any};
+  validateFn: (inputVal: CalendarValue) => { [key: string]: any };
```
Comment (vlio20) [inline src/app/date-picker/date-picker.component.ts]: please change it to the same alignment that was set here.

File: src/app/date-picker/date-picker.component.ts [hunk 4]
```diff
@@ -162,7 +162,7 @@ export class DatePickerComponent implements OnChanges,
 
   @HostListener('click')
   onClick() {
-    if (!this.isFocusedTrigger && !this.disabled) {
+    if (!this.isFocusedTrigger && !this.disabled && !this.config.apiControlOnly) {
```
Comment (vlio20) [inline src/app/date-picker/date-picker.component.ts]: please put this value in a variable - too much logic in one if statement.

File: src/app/date-picker/date-picker.component.ts [hunk 5]
```diff
@@ -311,6 +311,9 @@ export class DatePickerComponent implements OnChanges,
   }
 
   inputFocused() {
+    if (this.config.apiControlOnly) {
+      return;
+    }
```
Comment (vlio20) [inline src/app/date-picker/date-picker.component.ts]: please add blank line after if braces

File: src/app/date-picker/date-picker.service.ts [hunk 6]
```diff
@@ -16,6 +16,7 @@ export class DatePickerService {
     closeOnSelect: true,
     closeOnSelectDelay: 100,
     format: 'DD-MM-YYYY',
+    apiControlOnly: true,
```
Comment (vlio20) [inline src/app/date-picker/date-picker.service.ts]: the default should be backward compatible so please set it as false.

File: src/app/demo/demo/demo.component.html [hunk 7]
```diff
@@ -745,6 +745,25 @@ <h3 class="dp-options-section">Config options</h3>
         </div>
       </div>
 
+      <div class="dp-option" *ngIf="isValidConfig('apiControlOnly')">
+        <span class="dp-option-header">
+          Show picker via api control only (apiControlOnly):
```
Comment (vlio20) [inline src/app/demo/demo/demo.component.html]: After changing the configuration option, please don't forget to change the demo as well

File: src/app/date-picker/date-picker-config.model.ts [hunk 8]
```diff
@@ -8,6 +8,8 @@ export interface IDatePickerConfig extends IDayCalendarConfig,
                                            ITimeSelectConfig {
   closeOnSelect?: boolean;
   closeOnSelectDelay?: number;
+  openOnFocus?: boolean;
```
Comment (vlio20) [inline src/app/date-picker/date-picker-config.model.ts]: this should be available in all type of pickers, not only timeSelect

File: src/app/date-picker/date-picker.component.spec.ts [hunk 9]
```diff
@@ -1,9 +1,64 @@
 import {DatePickerComponent} from './date-picker.component';
+import {DatePickerService} from './date-picker.service';
+import {async, ComponentFixture, fakeAsync, TestBed, tick} from '@angular/core/testing';
+import {DebugElement, NO_ERRORS_SCHEMA} from '@angular/core';
+import {UtilsService} from '../common/services/utils/utils.service';
+import {DomHelper} from '../common/services/dom-appender/dom-appender.service';
+import {By} from '@angular/platform-browser';
+import {DayCalendarComponent} from '../day-calendar/day-calendar.component';
 
 describe('Component: DpDayPicker', () => {
```
Comment (vlio20) [inline src/app/date-picker/date-picker.component.spec.ts]: this type of test should be in e2e

File: src/app/date-picker/date-picker-config.model.ts (hunk 8 above)
Comment (piotr-dajlido) [reply src/app/date-picker/date-picker-config.model.ts]: This works for all pickers

File: src/app/date-picker/date-picker.component.spec.ts (hunk 9 above)
Comment (piotr-dajlido) [reply src/app/date-picker/date-picker.component.spec.ts]: I can't get protractor to trigger `focus` event, only `click` is supported. Also importing dependencies from my `spec` don't compile in e2e tests.

File: src/app/date-picker/date-picker-config.model.ts (hunk 8 above)
Comment (vlio20) [reply src/app/date-picker/date-picker-config.model.ts]: But it is odd that it is present only in the timeSelect

File: src/app/date-picker/date-picker.component.spec.ts (hunk 9 above)
Comment (vlio20) [reply src/app/date-picker/date-picker.component.spec.ts]: OK, lets add the click tests to e2e and I will think about something for the focus case

File: src/app/date-picker/date-picker-config.model.ts (hunk 8 above)
Comment (piotr-dajlido) [reply src/app/date-picker/date-picker-config.model.ts]: I don't understand what do you mean? This configuration as far as I am concerned is shared among other pickers, I have checked the demo page, and it worked perfectly.

**Comment by piotr-dajlido** [comment]
@vlio20 Hey, I have pushed necessary changes, is there anything else I can do ?

**Comment by vlio20** [comment]
@piotr-dajlido thanks a lot for you contribution!

**Comment by piotr-dajlido** [comment]
Sure, glad to contribute, when changes are expected to be published into npm ?

**Comment by vlio20** [comment]
Next few days

**Pull request merged**